### DISCUSSION TITLE
Best practices for implementing Idempotent Requests with FastAPI

### DISCUSSION THREAD

**Discussion by simon-liebehenschel** [fastapi/fastapi#3555] [Questions]
> API idempotency - safely retrying requests without accidentally performing the same operation twice. This is useful when an API call is disrupted in transit and you do not receive a response. For example, if a request does not respond due to a network connection error, you can retry the request with the same idempotency key to guarantee that no more than one charge is created.
> 
> To perform an idempotent request, provide an additional idempotency_key element to the request options.
> 
>  Idempotency works by saving the resulting status code and body of the first request made for any given idempotency key, regardless of whether it succeeded or failed. Subsequent requests with the same key return the same result, including 500 errors.
> 
> All POST requests accept idempotency keys. Sending idempotency keys in GET and DELETE requests has no effect and should be avoided, as these requests are idempotent by definition

I do not want to reinvent the wheel if there is already existing pythonic approach to implement Idempotent Requests for my API.

Otherwise, I think that adding to FastAPI routes put/get `idempotency_key` value inside my cache will be enough:

```python
@router.post("/send-email", status_code=status.HTTP_201_CREATED)
async def send_email(
        email: Email,
) -> CreateSmtpEmail:
    if response := await cache.get("idempotency_key"):
        return response
    
    # Continue if `idempotency_key` was not stored yet
```

**Comment by tomek360** [comment]
Did you try this library for fastapi?
https://pypi.org/project/fastapi-idempotent/

**Reply by ShahriyarR** [reply]
This one seems to be outdated as the last release was in Jun 22, 2022.
Instead found this one, but have not tried it yet: https://pypi.org/project/idemptx/

**Reply by RichieB2B** [reply]
I found a recent implementation that uses middleware, but have not tried It yet: https://github.com/buya25/fastapi-idempotency-key

**Comment by yudin-s** [comment]
The important part is that the idempotency key cannot be a single cache key like `"idempotency_key"`. It has to be scoped to the operation and the caller, otherwise two different users or two different endpoints can collide.

A production-ish shape is usually:

- read `Idempotency-Key` from a header, not from the body
- only require it for unsafe operations like `POST` / payment-like actions
- scope the storage key by user, method, path, and optionally a hash of the request body
- use an atomic lock/write in Redis or your DB so two identical requests racing at the same time do not both execute
- store the final status code and response body with a TTL
- if the same key is reused with a different body, return `409 Conflict`

Roughly:

```python
key = request.headers.get("Idempotency-Key")
body_hash = sha256(await request.body()).hexdigest()
cache_key = f"idem:{user.id}:{request.method}:{request.url.path}:{key}"

existing = await redis.get(cache_key)
if existing:
    saved = json.loads(existing)
    if saved["body_hash"] != body_hash:
        raise HTTPException(status_code=409, detail="Idempotency key reused with different request body")
    return Response(
        content=saved["body"],
        status_code=saved["status_code"],
        media_type=saved["media_type"],
    )
```

The tricky bit is the race before `existing` is written. For that, use something atomic like Redis `SET key value NX EX ...` for an in-progress marker, or a database table with a unique constraint on the idempotency key. Without that, two concurrent retries can both miss the cache and both send the email/charge.

So yes, caching the response is the right idea, but the safety comes from scoping + atomic claim + replaying the same stored response.

**Comment by violetta98** [comment]
Hello, I built [idemkit](https://github.com/idemkit/idemkit) after cleaning up duplicate charges one too many times. It runs your code once per key, even when two requests race or a worker dies.

The version everyone writes checks whether a key has been seen and replays the stored response. Two requests a millisecond apart both find nothing and both charge the card. And if the worker dies between charging and recording it, the retry charges again. Neither reproduces locally.

idemkit does it properly: an atomic claim instead of check-then-act, a lease that expires on the storage server's clock, and a fencing token so a stalled worker can't overwrite a good result.

```
from idemkit import idempotent, RedisBackend, MethodConfig 

@idempotent(
    backend=RedisBackend.from_url("redis://localhost:6379"), 
    config=MethodConfig(key_fields=["order_id"]), 
) 
async def charge(*, order_id, amount): 
    return await payments.charge(order_id, amount)
```

One core, three ways to use it: middleware for FastAPI/Flask/Django, a queue consumer wrapper or `@idempotent` on any function. Backends are Redis, Postgres, Mongo, DynamoDB, or in-memory for tests.

`pip install idemkit`, Apache-2.0

If you find it useful, please give it a try and let me know what you think!

**Comment by mmdverse** [comment]
I would keep this at the application boundary rather than trying to make it a generic FastAPI route cache. FastAPI can help read a header and return a response, but it cannot know whether an operation has actually happened.

The important change to the sketch in the question is that the key cannot be a constant cache key. At minimum it should be scoped by the authenticated subject, HTTP method, route, and a hash of the normalized request body. Otherwise two users, or two different operations from the same user, can replay each other.

For a database-backed API, the shape I usually reach for is a small idempotency table with a unique constraint on something like:

```text
(subject_id, method, route, idempotency_key)
```

The first request inserts a row in `processing` state inside the same transaction that creates the local business record. A concurrent request with the same key sees the unique conflict and either waits briefly / returns `409 still processing`, or replays the stored final response once the first request is complete. On completion, store the status code and a JSON response snapshot so retries can return the same result.

For the `send-email` example, there is one extra trap: a process can die after the SMTP provider accepted the email but before the API writes its final response. A simple cache cannot make that safe. I would create an outbox row with the same uniqueness constraint, commit it first, and let a worker send it. If the provider supports its own idempotency key, pass the same stable key there too. That turns the retry problem into a recoverable state transition instead of a race between "email sent" and "cache written".

So: header-based idempotency keys are a good interface, but the durable store and the side-effect boundary are where the guarantee actually lives. For simple CRUD, Redis with an atomic claim can be enough. For money, email, or webhooks, I would prefer a database uniqueness constraint plus an outbox pattern.
